fmt/flac.c

```diff
@@ -0,0 +1,343 @@
+}
+
+static FLAC__StreamDecoderWriteStatus on_write(const FLAC__StreamDecoder *decoder, const FLAC__Frame *frame, const FLAC__int32 *const buffer[], void *client_data) {
+	struct flac_file* flac_file = (struct flac_file*)client_data;
+
+	if (!flac_file->streaminfo.total_samples || !flac_file->streaminfo.channels)
+		return FLAC__STREAM_DECODER_WRITE_STATUS_ABORT;
+
+	if (flac_file->streaminfo.channels > 2)
+		return FLAC__STREAM_DECODER_WRITE_STATUS_ABORT;
+
+	if (!(flac_file->streaminfo.bits_per_sample == 8 || flac_file->streaminfo.bits_per_sample == 12 || flac_file->streaminfo.bits_per_sample == 16
+		|| flac_file->streaminfo.bits_per_sample == 20 || flac_file->streaminfo.bits_per_sample == 24 || flac_file->streaminfo.bits_per_sample == 32))
+		return FLAC__STREAM_DECODER_WRITE_STATUS_ABORT;
+
+	if (frame->header.number.sample_number == 0)
+	{
+		total_size = (size_t)(flac_file->streaminfo.total_samples * flac_file->streaminfo.channels * flac_file->streaminfo.bits_per_sample/8);
+		flac_file->uncomp_buf = (uint8_t*)malloc(total_size * ((flac_file->streaminfo.bits_per_sample == 8) ? sizeof(int8_t) : sizeof(int16_t)));
+		if (flac_file->uncomp_buf == NULL) {
+			log_appendf(4, "Error loading FLAC: Out of memory!");
+			return FLAC__STREAM_DECODER_WRITE_STATUS_ABORT;
+		}
+
+		samples_decoded = 0;
+	}
+
+	/* this isn't true, but it works fine for what we do */
+	uint32_t block_size = frame->header.blocksize*flac_file->streaminfo.channels;
+
+	const uint32_t samples_allocated = flac_file->streaminfo.total_samples*flac_file->streaminfo.channels;
+	if (samples_decoded+block_size > samples_allocated)
+		block_size = samples_allocated - samples_decoded;
+
+	uint32_t i = 0, j = 0;
+	switch (flac_file->streaminfo.bits_per_sample) {
+		case 8: {
+			int8_t *buf_ptr = flac_file->uncomp_buf + samples_decoded;
+			for (i = 0, j = 0; i < block_size; j++) {
+				buf_ptr[i++] = (int8_t)buffer[0][j];
+				if (flac_file->streaminfo.channels == 2)
+					buf_ptr[i++] = (int8_t)buffer[1][j];
+			}
+			break;
+		}
+		/* here we cheat by just increasing 12-bit to 16-bit */
+		case 12: case 16: {
+			int16_t *buf_ptr = (int16_t*)flac_file->uncomp_buf + samples_decoded;
+			uint32_t bit_shift = 16 - flac_file->streaminfo.bits_per_sample;
+			for (i = 0, j = 0; i < block_size; j++) {
+				buf_ptr[i++] = buffer[0][j] << bit_shift;
+				if (flac_file->streaminfo.channels == 2)
+					buf_ptr[i++] = buffer[1][j] << bit_shift;
+			}
+			break;
+		}
+		/* here we just make everything 16-bit, we don't support
+		   24-bit anyway */
+		case 20: case 24: case 32: {
+			int16_t *buf_ptr = (int16_t*)flac_file->uncomp_buf + samples_decoded;
+			uint32_t bit_shift = flac_file->streaminfo.bits_per_sample - 16;
+			for (i = 0, j = 0; i < block_size; j++) {
+				buf_ptr[i++] = buffer[0][j] >> bit_shift;
+				if (flac_file->streaminfo.channels == 2)
+					buf_ptr[i++] = buffer[1][j] >> bit_shift;
+			}
+			break;
+		}

```

Comment:
I don't see a way to combine these cases without (probably) degrading performance, so the duplication here is okay